### PR TITLE
(#Closes 1575 and 2248) Change to line length module to avoid line continuation before any code

### PR DESCRIPTION
**Pull request by LonelyCat124**
I think this fixes the behaviour seen in #1575 and #2248. The fix is simple, which just prevents line continuation before the first statement on a line.

I'm not clear how to have a good test for this, I'll put one together but any improvement ideas would be useful.

### CODE REVIEW

**Comment by LonelyCat124** [comment]
Ok, test added, I'm ready for review for this.

**Comment by LonelyCat124** [comment]
The only thing I can think of is if you had a statement that was already indented too far this wouldn't work, but there is no good solution to that problem I think.

File: src/psyclone/line_length.py [hunk 1]
```diff
@@ -13,9 +13,11 @@
 
 def find_break_point(line, max_index, key_list):
     ''' find the most appropriate break point for a fortran line '''
-
```
Comment (arporter) [inline src/psyclone/line_length.py:16]: Unfortunately (for you) this is ancient code. Please could you change the copyright header to be the one we now use and document the arguments and return value/type of this method.
Please could you also remove the `from __future__` import near the top.

File: src/psyclone/tests/line_length_test.py [hunk 2]
```diff
@@ -330,6 +330,32 @@ def test_edge_conditions_comments():
     assert output_string == expected_output
 
 
+def test_lone_lines_allocate(fortran_reader, fortran_writer):
```
Comment (arporter) [inline src/psyclone/tests/line_length_test.py:333]: You could keep this test but I would suggest testing the method `find_break_point` directly - you can then give it a number of different strings by e.g. parameterising the test (pytest parametrize).

File: src/psyclone/tests/line_length_test.py [hunk 3]
```diff
@@ -330,6 +330,32 @@ def test_edge_conditions_comments():
...
+    IF (.NOT. ALLOCATED(pressure_prsc)) &
+      ALLOCATE(pressure_prsc        ( some_type%longstringname1, &
+                                      some_type%longstringname2, &
+                                      some_type%longstringname3, &
+                                      some_type%longstringname4 ))
+    end subroutine test
+    '''
+
+    psyir = fortran_reader.psyir_from_source(code)
+    from psyclone.line_length import FortLineLength
```
Comment (arporter) [inline src/psyclone/tests/line_length_test.py:348]: Please move this import to the top.

File: src/psyclone/line_length.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/line_length.py:16]: Please could you also move the `import re` at L53 to the top (pylint complains).

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for fixing this @LonelyCat124. The method you've touched unfortunately needs bringing up-to-date with our current practises. Aside from that, I think it would be good (and hopefully simple) to just test that method directly for a variety of input strings.
I've triggered the integration tests so will check those next time.
This is a bug fix so no doc update is required but I'll check that the Ref Guide builds OK once you've improved the doc string of the method.

File: src/psyclone/line_length.py (hunk 1 above)
Comment (LonelyCat124) [reply src/psyclone/line_length.py:16]: Updated.
There is no `from __future__` import in my version of the file I can find.

Comment (LonelyCat124) [reply src/psyclone/line_length.py:16]: Ah future is in the tests - i removed it there.

File: src/psyclone/tests/line_length_test.py (hunk 3 above)
Comment (LonelyCat124) [reply src/psyclone/tests/line_length_test.py:348]: Done - was already imported.

File: src/psyclone/tests/line_length_test.py (hunk 2 above)
Comment (LonelyCat124) [reply src/psyclone/tests/line_length_test.py:333]: Done

**Comment by LonelyCat124** [comment]
@arporter I hope I've addressed the comments correctly, pycodestyle and pylint all look good on my end now, and I believe the unit tests fully cover the functionality.

File: src/psyclone/line_length.py [hunk 4]
```diff
@@ -1,29 +1,78 @@
...
+
+    :param str line: the Fortran code string to find the line break point for.
+    :param int max_index: the maximum index in line to search for the line
+                          break point.
+    :param key_list: list of potential symbols to break the line at. The
+                     members of the list early in the ordering have priority
+                     for breaking the line, i.e. if the list contains multiple
+                     elements, any possible position of the first element will
+                     be found before trying any other element of the list.
+    :type key_list: list of str.
```
Comment (arporter) [inline src/psyclone/line_length.py:59]: `List[str]` please.

File: src/psyclone/line_length.py [hunk 5]
```diff
@@ -1,29 +1,78 @@
...
+
+    :returns: index to break the line into multiple lines.
+    :rtype: int
+
+    :raises Exception: if no suitable break point is found in line.
```
Comment (arporter) [inline src/psyclone/line_length.py:64]: s/Exception/InternalError/

File: src/psyclone/line_length.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/line_length.py:16]: Oops, my mistake and thanks.

**Review by arporter — CHANGES_REQUESTED** [review]
Almost perfect now.
Just a couple of doc string fixes needed.
Ref. guide builds fine.
Integration tests are fine.
Changes are pylint clean.
Test file covers source file 100%.
Once those changes are made this can go on.

File: src/psyclone/line_length.py (hunk 4 above)
Comment (LonelyCat124) [reply src/psyclone/line_length.py:59]: Done

File: src/psyclone/line_length.py (hunk 5 above)
Comment (LonelyCat124) [reply src/psyclone/line_length.py:64]: Done

**Review by arporter — APPROVED** [review]
All requested changes have been made (thanks).
Will proceed to merge.

**Pull request merged**